src/btf/compat.cpp

```diff
@@ -0,0 +1,116 @@
+#include "btf/compat.h"
+#include "struct.h"
+
+namespace bpftrace::btf {
+
+char CompatTypeError::ID;
+
+void CompatTypeError::log(llvm::raw_ostream &OS) const
+{
+  OS << msg_;
+}
+
+Result<SizedType> getCompatType([[maybe_unused]] const Void &type)
+{
+  return CreateVoid();
+}
+
+Result<SizedType> getCompatType(const Integer &type)
+{
+  return CreateInt(8 * type.bytes());
+}
+
+Result<SizedType> getCompatType(const Pointer &type)
+{
+  auto t = type.element_type();
+  if (!t) {
+    return t.takeError();
+  }
+  auto ty = getCompatType(*t);
+  if (!ty) {
+    return ty.takeError();
+  }
+  return CreatePointer(*ty);
+}
+
+Result<SizedType> getCompatType(const Array &type)
+{
+  auto t = type.element_type();
+  if (!t) {
+    return t.takeError();
+  }
+  auto ty = getCompatType(*t);
+  if (!ty) {
+    return ty.takeError();
+  }
+  return CreateArray(type.element_count(), *ty);
+}
+
+Result<SizedType> getCompatType(const Struct &type)
+{
+  std::vector<SizedType> fields;
+  std::vector<std::string_view> idents;
+  auto f = type.fields();
+  if (!f) {
+    return f.takeError();
+  }
+  // We have no ability to control how things are packed,
+  // so this may or may not generate the correct type.
```

Comment:
oof 😢

Reply:
The packing of `SizedType` is an implementation detail, although it *should* align with the LLVM layout. So they *should* be the same. The packing of the BTF type is part of the representation, but we have no ability to pass that through when creating the `SizedType` here.

(Obviously removing the load-bearing "should"s there is part of the motivation for eventually aligning on a single consistent type representation with BTF)